### DISCUSSION TITLE
Frontend에서 Backend(MLFuncs)의 uploadfile API에 접근 시 발생한 CORS 에러

### DISCUSSION THREAD

**Discussion by seahahn** [AI-Play/Front-Web#21] [General]
[CORS 관련 링크](https://developer.mozilla.org/ko/docs/Web/HTTP/CORS)

프론트에서 백앤드에 csv 파일을 전송하여 데이터프레임화 된 JSON을 받아오는 uploadfile API에 접근하였음.
그 결과 아래와 같은 에러 발생.
포트 8000이 백앤드, 3000이 프론트앤드임.

Access to fetch at 'http://localhost:8000/uploadfile' from origin 'http://localhost:3000' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.

교차 출처 리소스 공유 (CORS)라는 것으로 인해 접근이 제한되어 에러가 발생한 것.
임시로 프론트에서 cors mode를 no-cors로 설정하여 진행 중.
그러나 기능 구현이 다 되고 PaaS에 배포할 시점이 되면 그때는 서버 측에서 Access-Control-Allow-Origin 헤더를 설정하여 CORS 문제를 해결해야 함.

**Comment by seahahn** [comment]
# 삽질 기록

## 1. React Component의 html 작업 중 form - input에서 파일 업로드가 되지 않는 문제(라고 착각)

### 문제 설명

```javascript
/* 불필요한 부분은 생략하였음 */
function DataUpload() {

  const [file, setFile] = useState(); // 파일을 저장하는 State

  const handleChange = event => {
    setFile(event.target.files[0]); // file input에 파일 지정하면 file State에 들어가게 함
    console.log(file); // 원래 같으면 여기서 file에 잘 들어가신 파일 정보가 잘 나오셔야 함.
  };

  return (
    <div>
      <form onSubmit={handleSubmit}>
        <input type="file" accept={UPLOAD_ACCEPT} onChange={handleChange}/> // 여기서 파일 지정하면 file State에 들어가야 함
      </form>
    </div>
  )
}
```
그런데 setFile을 하고 나면 file을 console.log로 찍었을 때 업로드한 파일 정보가 나와야 할 것만 같은데 나오질 않음.
웃기는건 새로운 파일을 업로드하면 console.log에는 그 전에 업로드했던 파일의 정보가 나옴.
마치 한 스텝 느리게 따라오는 듯하여 굉장히 약이 올랐으나, 온갖 레퍼런스를 참고해서 지지고 볶아봐도 고쳐지질 않음.

### 해결 방법
해결 방법이랄 것도 없었음. 그냥 로그만 바로 안 나올 뿐이지 파일은 정상적으로 file State에 들어가고 있었음.
이것때문에 시간을... 뭐 아무튼, setFile 바로 뒤에 log를 찍은게 undefined가 나오더라도 파일 업로드는 된다.
정 업로드 여부를 확인하고 싶다면, 별도로 file의 정보를 포함한 alert를 띄울 버튼을 따로 만든 후에 파일 업로드 후 버튼을 눌러서 제대로 올라갔는지 볼 수도 있다.

```javascript
// 이렇게 만들어서 올라갔는지 확인했음
<button type="button" onClick={()=>alert(file.name)}>Alert</button>
```

**Reply by seahahn** [reply]
## 2. CORS 설정으로 인해 백앤드로 파일 전송을 하지 못함

### 문제 설명

1) 제목 그대로 CORS 문제로 인해 프론트앤드에서 백앤드의 API를 사용하질 못했음.
같은 localhost이지만 포트 번호가 달랐음. 프론트는 3000, 백앤드는 8000.

2) 프론트앤드에서 fetch로 API 요청 시 CORS의 mode를 no-cors로 했었는데, 이때 'Unexpected end of input' 에러가 발생함.
참고 링크 - [Handle response - SyntaxError: Unexpected end of input when using mode: 'no-cors'](https://stackoverflow.com/questions/43317967/handle-response-syntaxerror-unexpected-end-of-input-when-using-mode-no-cors)

3) header를 아래와 같이 설정했는데, 이걸 눈치를 못 채고 또 시간을 날려먹음.
```javascript
    const config = {
      method: "POST",
      mode: CORS,
      Accept: UPLOAD_ACCEPT,
      headers: {
        'Content-Type': 'multipart/form-data', // 문제 원인
      },
      body: formData,
    };
```

### 해결 방법

1) 백앤드 쪽에 아래와 같이 CORS 설정을 추가하였음.

#### MLFuncs - main.py

```python
from fastapi.middleware.cors import CORSMiddleware

origins = [
    "http://localhost:3000" # 포트 지정 안 하면 CORS 에러 발생
]

app.add_middleware(
    CORSMiddleware,
    allow_origins=origins,
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)
```

2) no-cors를 cors로 변경하니 해결됨.

3) 헤더를 아예 빼버리니 정상 작동함.
아래 링크의 내용을 참고하여 headers를 추가한 것인데, 예제가 당연히 작동할 것이라 생각해서 이게 문제임을 빨리 알지 못하였음.
[React File Upload](https://www.filestack.com/fileschool/react/react-file-upload/)
정확히는 모르겠으나, 예제는 Axios를 사용하므로 설정이 다른건가 싶음.

---

<참고 자료>
- [CORS (Cross-Origin Resource Sharing) - FastAPI](https://fastapi.tiangolo.com/tutorial/cors/)
- [Handle response - SyntaxError: Unexpected end of input when using mode: 'no-cors'](https://stackoverflow.com/questions/43317967/handle-response-syntaxerror-unexpected-end-of-input-when-using-mode-no-cors)
- [[React]CORS 에러 해결](https://developer-talk.tistory.com/91)
- [Uploading a file - MDN](https://developer.mozilla.org/ko/docs/Web/API/Fetch_API/Using_Fetch#uploading_a_file)
- [Uploading Files with React.js](https://www.pluralsight.com/guides/uploading-files-with-reactjs)

**Reply by KayyoungHL** [reply]
FastAPI 기본 포트가 8000번인데, 이걸 3000으로 바꾼다는 말씀이시죠?
뭔가 접근하는 쪽에서 접근하려는 주소를 바꿔야 한다니 뭔가 상황이 조금 이상하네요 ㅋㅋ

**Reply by seahahn** [reply]
> FastAPI 기본 포트가 8000번인데, 이걸 3000으로 바꾼다는 말씀이시죠? 뭔가 접근하는 쪽에서 접근하려는 주소를 바꿔야 한다니 뭔가 상황이 조금 이상하네요 ㅋㅋ

아뇨, 서버 포트를 바꾸는게 아니라 origin에서 접근 허용할 도메인과 포트를 지정해주는 거에요.
위에 origin에 localhost:3000 지정한 것은 localhost:3000의 접근을 허용한다는 의미입니다.

**Reply by KayyoungHL** [reply]
아하 이해했습니다.
FastAPI에서 접근 허용하는 서버를 0.0.0.0:0 으로 해줘야 하나 보네요.
일단 경호님께서 작성하신 코드를 머지하고 나중에 확인해보겠습니다.
